lib/datadog/di/probe_notification_builder.rb

```diff
@@ -0,0 +1,216 @@
+            timestamp: timestamp,
+            evaluationErrors: [],
+            probe: {
+              id: probe.id,
+              version: 0,
+              location: location,
+            },
+            language: 'ruby',
+            # TODO add test coverage for callers being nil
+            stack: stack,
+            captures: captures,
+          },
+          # In python tracer duration is under debugger.snapshot,
+          # but UI appears to expect it here at top level.
+          duration: duration ? (duration * 10**9).to_i : nil,
+          host: nil,
+          logger: {
+            name: probe.file,
+            method: probe.method_name || 'no_method',
+            thread_name: Thread.current.name,
+            thread_id: thread_id,
+            version: 2,
+          },
+          "dd.trace_id": 136035165280417366521542318182735500431,
+          "dd.span_id": 17576285113343575026,
+          ddsource: 'dd_debugger',
+          message: probe.template && evaluate_template(probe.template,
+            duration: duration ? duration * 1000 : nil),
+          timestamp: timestamp,
+        }
+      end
+
+      def build_status(probe, message:, status:)
+        {
+          service: settings.service,
+          timestamp: timestamp_now,
+          message: message,
+          ddsource: 'dd_debugger',
+          debugger: {
+            diagnostics: {
+              probeId: probe.id,
+              probeVersion: 0,
+              runtimeId: Core::Environment::Identity.id,
+              parentId: nil,
+              status: status,
+            },
+          },
+        }
+      end
+
+      def format_callers(callers)
+        callers.map do |caller|
+          if caller =~ /\A([^:]+):(\d+):in `([^']+)'\z/
+            {
+              fileName: $1, function: $3, lineNumber: Integer($2),
+            }
+          else
+            {
+              fileName: 'unknown', function: 'unknown', lineNumber: 0,
+            }
+          end
+        end
+      end
+
+      def evaluate_template(template, **vars)
+        message = template.dup
+        vars.each do |key, value|
+          message.gsub!("{@#{key}}", value.to_s)
+        end
+        message
+      end
+
+      def timestamp_now
+        (Time.now.to_f * 1000).to_i
+      end
+
+      def get_local_variables(trace_point)
+        # binding appears to be constructed on access, therefore
+        # 1) we should attempt to cache it and
+        # 2) we should not call +binding+ until we actually need variable values.
+        binding = trace_point.binding
+
+        # steep hack - should never happen
+        return {} unless binding
+
+        binding.local_variables.each_with_object({}) do |name, map|
+          value = binding.local_variable_get(name)
+          map[name] = value
+        end
+      end
+
+      def thread_id
+        thread = Thread.current
+        if thread.respond_to?(:native_thread_id)
+          # Ruby 3.1+
+          thread.native_thread_id
```

Comment:
Be careful with using `native_thread_id` here -- in Ruby 3.3+ ruby threads can move across OS threads (see [M:N announcement](https://www.ruby-lang.org/en/news/2023/12/25/ruby-3-3-0-released/#:~:text=of%20RubyKaigi.-,M%3AN%20thread%20scheduler,-M%3AN%20thread)). This behavior change is off by default in the main ractor but that may change...

Reply:
I don't use it myself for anything and I made a note to inquire with my team what it is supposed to be.

Reply:
I removed thread_id sending because currently DI does not need it.